tsfc/coffee_mode.py

```diff
@@ -0,0 +1,317 @@
+    factor_group = OrderedDict()
+    for monomial in monomial_sum:
+        for sum_indices, oa in optimal_atomics:
+            if frozenset(monomial.sum_indices) == frozenset(sum_indices) and oa in monomial.atomics:
+                # Add monomial to the list of corresponding optimal atomic
+                factor_group.setdefault((sum_indices, oa), []).append(monomial)
+                break
+        else:
+            # Add monomials that do no have argument factors to new MonomialSum
+            new_monomial_sum.add(monomial.sum_indices, monomial.atomics, monomial.rest)
+    # We should not drop monomials
+    assert sum(map(len, itervalues(factor_group))) + len(new_monomial_sum) == len(monomial_sum)
+
+    for (sum_indices, oa), monomials in iteritems(factor_group):
+        if len(monomials) == 1:
+            # Just one monomial with this group, add to new MonomialSum straightaway
+            monomial, = monomials
+            new_monomial_sum.add(monomial.sum_indices, monomial.atomics, monomial.rest)
+            continue
+        all_atomics = []  # collect all atomics from monomials
+        all_rest = []  # collect all rest from monomials
+        for monomial in monomials:
+            _atomics = list(monomial.atomics)
+            _atomics.remove(oa)  # remove common factor
+            all_atomics.append(_atomics)
+            all_rest.append(monomial.rest)
+        # Create new MonomialSum for the factorised out terms
+        sub_monomial_sum = MonomialSum()
+        for _atomics, _rest in zip(all_atomics, all_rest):
+            sub_monomial_sum.add((), _atomics, _rest)
+        sub_monomial_sum = optimise_monomial_sum(sub_monomial_sum, argument_indices)
+        assert len(sub_monomial_sum) > 0
+        if len(sub_monomial_sum) == 1:
+            # result is a product, add to new MonomialSum directly
+            sub_monomial, = sub_monomial_sum
+            new_atomics = sub_monomial.atomics
+            new_atomics += (oa,)  # add back common factor
+            new_rest = sub_monomial.rest
+        else:
+            # result is a sum, we need to create new node
+            new_node = monomial_sum_to_expression(sub_monomial_sum)
+            new_atomics = [oa]
+            new_rest = one
+            if set(argument_indices) & set(new_node.free_indices):
+                new_atomics.append(new_node)
+            else:
+                new_rest = new_node
+        new_monomial_sum.add(sum_indices, new_atomics, new_rest)
+    return new_monomial_sum
+
+
+def optimise_monomial_sum(monomial_sum, argument_indices):
+    """Choose optimal common atomic subexpressions and factorise a
+    :class:`MonomialSum` object.
+
+    :arg monomial_sum: a :class:`MonomialSum` object
+    :arg argument_indices: tuple of argument indices
+
+    :returns: factorised `MonomialSum` object
+    """
+    all_optimal_atomics = []  # [(sum_indces, optimal_atomics)]
+    for sum_indices in unique_sum_indices(monomial_sum):
```

Comment:
OK, so I see what you actually do is split the monomial sum into groups along the sum indices (simple, right?), but you do it in an absolutely horrible and convoluted manner.

First, there is this `unique_sum_indices` function which generates the keys for the split. That function is only used here, which in itself should make you consider inlining it. (Remember the rule: used only once => _consider_ inlining)

Next you pass the whole monomial sum to `find_optimal_atomics` along with one particular split key from `unique_sum_indices`. Then you do a filtering of the whole monomial sum with that key at the beginning of `find_optimal_atomics`. So in this loop you create a mapping (`all_optimal_atomics`) from split keys to `optimal_atomics` (whatever that means). (Yes, you represent it as list of pairs rather than, let's say, a `dict`, but it essentially represents a mapping.)

Finally, you pass the whole monomial sum again, along with the mapping produced above to `factorise_atomics`. Then `factorise_atomics` begins with what? -- You guessed it! -- Iterating over the whole monomial sum and the mapping from split keys to "optimal atomics" in a nested pair of for loops and filtering out the matches, to produce something called `factor_group` which is what the rest of that function actually operates on.

There is no need to scatter among three functions something as simple as splitting a list of items into groups, meanwhile unnecessarily complicating APIs. Just write something like this:
```python
groups = OrderedDict()
for monomial in monomial_sum:
    key = frozenset(monomial.sum_indices)
    groups.setdefault(key, MonomialSum()).add(*monomial)

factor_groups = []
for key_sum_indices, group_monomial_sum in iteritems(groups):
    # No more filtering needed in find_optimal_atomics,
    # maybe also pass key_sum_indices if really needed.
    optimal_atomics = find_optimal_atomics(group_monomial_sum, argument_indices)
    factor_groups.append((group_monomial_sum, key_sum_indices))

# No more cross-matching needed in factorise_atomics
return factorise_atomics(factor_groups, argument_indices)
```